url29ae058134_yuriy_chumak_jatha_tgz-pJ8-org_jatha_LispJ8/src/org/jatha/read/LispParser.java

```diff
@@ -0,0 +1,1065 @@
+				if (isTerminator(ch))       // Terminate reading token.
+				{
+					unread(ch);
+					parseState = READING_NOTHING;
+					return tokenToLispValue(
+							f_caseSensitivity == UPCASE   ? token.toString().toUpperCase() :
+							f_caseSensitivity == DOWNCASE ? token.toString().toLowerCase() :
+								token.toString()
+							);
+				}
+				
+				token.append(ch);
+				break;
+
+			case READING_MIXED_CASE_SYMBOL:
+				if (isOrBar(ch))       // Terminate reading token.
+				{
+					String s = token.toString();
+
+					token.append(ch);
+					parseState = READING_NOTHING;
+					// Strip off the beginning and ending Or Bars.
+					return tokenToLispValue(s.substring(1, s.length()));
+				}
+				
+				token.append(ch);
+				break;
+
+			case READING_STRING:
+				if (ch == BACKSLASH)  // Next char is always in the string
+				{
+					try { intCh = inputReader.read(); }
+					catch (IOException ioex) { break; }
+					if (intCh < 0)
+						throw new EOFException("Premature end of LISP input at: " + token.toString());
+					
+					ch = (char)intCh;
+					token.append(ch);
+					break;
+				}
+
+				if (ch == DOUBLE_QUOTE)
+				{
+					token.append(ch);
+					parseState = READING_NOTHING;
+					return tokenToLispValue(token.toString());
+				}
+				
+				token.append(ch);
+				break;
+			case READING_NOTHING:
+				if (!isSpace(ch))         /* Start reading a token */
+				{
+					if (isLparen(ch))
+						return read_list_token(inputReader);
+					if (isRparen(ch))
+					{
+						System.err.println("WARNING: Too many right parentheses.  NIL assumed.");
+						return NIL;
+					}
+					if (isQuote(ch))
+						return read_quoted_token(inputReader, UPCASE);
+					if (isColon(ch))
+						return read_quoted_token(inputReader, PRESERVE);
+					if (isDoubleQuote(ch)) {
+						token.append(ch);
+						parseState = READING_STRING;
+						break;
+					}
+					if (isPound(ch))
+						return(applyReaderMacro(inputReader));
+					if (isBackQuote(ch))
+						return(read_backquoted_token(inputReader));
+					if (isComma(ch))
+						return(read_comma_token(inputReader));
+					
+					if (isOrBar(ch)) {
+						token.append(ch);
+						parseState = READING_MIXED_CASE_SYMBOL;
+						break;
+					}
+					
+					parseState = READING_SYMBOL;
+					unread(ch);
+				}  /* if (!isSpace(ch)) */
+				break;
+			}
+		} /* main WHILE loop */
+
+		/* WE ONLY EXECUTE THIS CODE IF WE HIT end of input string or file. */
+		if (token.length() > 0)
+			return(tokenToLispValue(token.toString()));
+		else
+			return(NIL);
+	}
+	
+	private final void unread(char ch)
+	{
+		try { inputReader.unread(ch); }
+		catch (IOException ioex) {
+			System.err.println("\n *** I/O error while unreading character '" + ch + "'.");
+		}
+	}
+
+
+  /**
+   * Reads one list expression from the input stream and returns it.
+   * The input pointer should be on the character following the left parenthesis.
+   */
+  public LispValue read_list_token(PushbackReader stream) throws EOFException
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.